File: app/src/main/java/org/michaelbel/application/ui/view/RatingView.java
package org.michaelbel.application.ui.view;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.support.annotation.IntDef;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.AppCompatImageView;
import android.widget.LinearLayout;

import org.michaelbel.application.R;
import org.michaelbel.application.moviemade.LayoutHelper;
import org.michaelbel.application.moviemade.Theme;

@SuppressWarnings("all")
public class RatingView extends LinearLayout {

    private StarView[] stars = new StarView[5];

    public RatingView(Context context) {
        super(context);

        setOrientation(HORIZONTAL);

        int starCount = 5;
        for (int i = 0; i < starCount; i++) {
            StarView view = new StarView(context);
            view.setLayoutParams(LayoutHelper.makeLinear(20, 20));
            addView(view);
            stars[i] = view;
        }
    }

    public void setRating(float rating) {
        float myRating = rating / 2;

        int a = (int) myRating;
        int b = (int) (10 * (myRating - a));
        int c;
        c = (b >= 5) ? 5 : 0;

        if (c == 0) {
            switch (a) {
                case 0:
                    for(StarView star : stars) {
                        star.setIcon(StarView.ICON_STAR_BORDER);
                    }
                    break;
                case 1:
                    stars[0].setIcon(StarView.ICON_STAR);
                    stars[1].setIcon(StarView.ICON_STAR_BORDER);
                    stars[2].setIcon(StarView.ICON_STAR_BORDER);
                    stars[3].setIcon(StarView.ICON_STAR_BORDER);
                    stars[4].setIcon(StarView.ICON_STAR_BORDER);
                    break;
                case 2:
                    stars[0].setIcon(StarView.ICON_STAR);
                    stars[1].setIcon(StarView.ICON_STAR);
                    stars[2].setIcon(StarView.ICON_STAR_BORDER);
                    stars[3].setIcon(StarView.ICON_STAR_BORDER);
                    stars[4].setIcon(StarView.ICON_STAR_BORDER);
                    break;
                case 3:
                    stars[0].setIcon(StarView.ICON_STAR);
                    stars[1].setIcon(StarView.ICON_STAR);
                    stars[2].setIcon(StarView.ICON_STAR);
                    stars[3].setIcon(StarView.ICON_STAR_BORDER);
                    stars[4].setIcon(StarView.ICON_STAR_BORDER);
                    break;
                case 4:
                    stars[0].setIcon(StarView.ICON_STAR);
                    stars[1].setIcon(StarView.ICON_STAR);
                    stars[2].setIcon(StarView.ICON_STAR);
                    stars[3].setIcon(StarView.ICON_STAR);
                    stars[4].setIcon(StarView.ICON_STAR_BORDER);
                    break;
                case 5:
                    for(StarView star : stars) {
                        star.setIcon(StarView.ICON_STAR);
                    }
                    break;
            }
        } else if (c == 5) {
            switch (a) {
                case 0:
                    stars[0].setIcon(StarView.ICON_STAR_HALF);
                    stars[1].setIcon(StarView.ICON_STAR_BORDER);
                    stars[2].setIcon(StarView.ICON_STAR_BORDER);
                    stars[3].setIcon(StarView.ICON_STAR_BORDER);
                    stars[4].setIcon(StarView.ICON_STAR_BORDER);
                    break;
                case 1:
                    stars[0].setIcon(StarView.ICON_STAR);
                    stars[1].setIcon(StarView.ICON_STAR_HALF);
                    stars[2].setIcon(StarView.ICON_STAR_BORDER);
                    stars[3].setIcon(StarView.ICON_STAR_BORDER);
                    stars[4].setIcon(StarView.ICON_STAR_BORDER);
                    break;
                case 2:
                    stars[0].setIcon(StarView.ICON_STAR);
                    stars[1].setIcon(StarView.ICON_STAR);
                    stars[2].setIcon(StarView.ICON_STAR_HALF);
                    stars[3].setIcon(StarView.ICON_STAR_BORDER);
                    stars[4].setIcon(StarView.ICON_STAR_BORDER);
                    break;
                case 3:
                    stars[0].setIcon(StarView.ICON_STAR);
                    stars[1].setIcon(StarView.ICON_STAR);
                    stars[2].setIcon(StarView.ICON_STAR);
                    stars[3].setIcon(StarView.ICON_STAR_HALF);
                    stars[4].setIcon(StarView.ICON_STAR_BORDER);
                    break;
                case 4:
                    stars[0].setIcon(StarView.ICON_STAR);
                    stars[1].setIcon(StarView.ICON_STAR);
                    stars[2].setIcon(StarView.ICON_STAR);
                    stars[3].setIcon(StarView.ICON_STAR);
                    stars[4].setIcon(StarView.ICON_STAR_HALF);
                    break;
            }
        }
    }

    public static class StarView extends AppCompatImageView {

        public static final int ICON_STAR = 0;
        public static final int ICON_STAR_HALF = 1;
        public static final int ICON_STAR_BORDER = 2;

        @IntDef({
                ICON_STAR,
                ICON_STAR_HALF,
                ICON_STAR_BORDER
        })
        private @interface Style {}

        private Drawable icon;

        public StarView(Context context) {
            super(context);
        }

        public void setIcon(@Style int style) {
            if (style == ICON_STAR) {
                icon = Theme.getIcon(R.drawable.ic_star, ContextCompat.getColor(getContext(), Theme.accentColor()));
            } else if (style == ICON_STAR_HALF) {
                icon = Theme.getIcon(R.drawable.ic_star_half, ContextCompat.getColor(getContext(), Theme.accentColor()));
            } else if (style == ICON_STAR_BORDER) {
                icon = Theme.getIcon(R.drawable.ic_star_border, ContextCompat.getColor(getContext(), Theme.accentColor()));
            }

            setImageDrawable(icon);
        }
    }
}

File: app/src/main/java/org/michaelbel/moviemade/presentation/common/RatingView.kt
package org.michaelbel.moviemade.presentation.common

import android.content.Context
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.widget.LinearLayout
import androidx.appcompat.widget.AppCompatImageView
import androidx.appcompat.widget.LinearLayoutCompat
import org.michaelbel.moviemade.R
import org.michaelbel.moviemade.ktx.getIcon
import org.michaelbel.moviemade.ktx.toDp

class RatingView(context: Context, attrs: AttributeSet?): LinearLayoutCompat(context, attrs) {

    companion object {
        const val ICON_STAR = 0
        const val ICON_STAR_HALF = 1
        const val ICON_STAR_BORDER = 2
    }

    private val stars = arrayOfNulls<StarView>(5)

    init {
        orientation = HORIZONTAL

        val starCount = 5
        for (i in 0 until starCount) {
            val view = StarView(context)
            view.layoutParams = LinearLayout.LayoutParams(20F.toDp(context), 20F.toDp(context))
            addView(view)
            stars[i] = view
        }
    }

    fun setRating(rating: Float) {
        val myRating = rating / 2

        val a = myRating.toInt()
        val b = (10 * (myRating - a)).toInt()
        val c: Int
        c = if (b >= 5) 5 else 0

        if (c == 0) {
            when (a) {
                0 -> for (star in stars) {
                    star?.setIcon(ICON_STAR_BORDER)
                }
                1 -> {
                    stars[0]?.setIcon(ICON_STAR)
                    stars[1]?.setIcon(ICON_STAR_BORDER)
                    stars[2]?.setIcon(ICON_STAR_BORDER)
                    stars[3]?.setIcon(ICON_STAR_BORDER)
                    stars[4]?.setIcon(ICON_STAR_BORDER)
                }
                2 -> {
                    stars[0]?.setIcon(ICON_STAR)
                    stars[1]?.setIcon(ICON_STAR)
                    stars[2]?.setIcon(ICON_STAR_BORDER)
                    stars[3]?.setIcon(ICON_STAR_BORDER)
                    stars[4]?.setIcon(ICON_STAR_BORDER)
                }
                3 -> {
                    stars[0]?.setIcon(ICON_STAR)
                    stars[1]?.setIcon(ICON_STAR)
                    stars[2]?.setIcon(ICON_STAR)
                    stars[3]?.setIcon(ICON_STAR_BORDER)
                    stars[4]?.setIcon(ICON_STAR_BORDER)
                }
                4 -> {
                    stars[0]?.setIcon(ICON_STAR)
                    stars[1]?.setIcon(ICON_STAR)
                    stars[2]?.setIcon(ICON_STAR)
                    stars[3]?.setIcon(ICON_STAR)
                    stars[4]?.setIcon(ICON_STAR_BORDER)
                }
                5 -> for (star in stars) {
                    star?.setIcon(ICON_STAR)
                }
            }
        } else {
            when (a) {
                0 -> {
                    stars[0]?.setIcon(ICON_STAR_HALF)
                    stars[1]?.setIcon(ICON_STAR_BORDER)
                    stars[2]?.setIcon(ICON_STAR_BORDER)
                    stars[3]?.setIcon(ICON_STAR_BORDER)
                    stars[4]?.setIcon(ICON_STAR_BORDER)
                }
                1 -> {
                    stars[0]?.setIcon(ICON_STAR)
                    stars[1]?.setIcon(ICON_STAR_HALF)
                    stars[2]?.setIcon(ICON_STAR_BORDER)
                    stars[3]?.setIcon(ICON_STAR_BORDER)
                    stars[4]?.setIcon(ICON_STAR_BORDER)
                }
                2 -> {
                    stars[0]?.setIcon(ICON_STAR)
                    stars[1]?.setIcon(ICON_STAR)
                    stars[2]?.setIcon(ICON_STAR_HALF)
                    stars[3]?.setIcon(ICON_STAR_BORDER)
                    stars[4]?.setIcon(ICON_STAR_BORDER)
                }
                3 -> {
                    stars[0]?.setIcon(ICON_STAR)
                    stars[1]?.setIcon(ICON_STAR)
                    stars[2]?.setIcon(ICON_STAR)
                    stars[3]?.setIcon(ICON_STAR_HALF)
                    stars[4]?.setIcon(ICON_STAR_BORDER)
                }
                4 -> {
                    stars[0]?.setIcon(ICON_STAR)
                    stars[1]?.setIcon(ICON_STAR)
                    stars[2]?.setIcon(ICON_STAR)
                    stars[3]?.setIcon(ICON_STAR)
                    stars[4]?.setIcon(ICON_STAR_HALF)
                }
            }
        }
    }

    private inner class StarView(context: Context): AppCompatImageView(context) {

        private var icon: Drawable? = null

        fun setIcon(style: Int) {
            when (style) {
                ICON_STAR -> icon = getIcon(R.drawable.ic_star, R.color.accent_yellow)
                ICON_STAR_HALF -> icon = getIcon(R.drawable.ic_star_half, R.color.accent_yellow)
                ICON_STAR_BORDER -> icon = getIcon(R.drawable.ic_star_border, R.color.accent_yellow)
            }

            setImageDrawable(icon)
        }
    }
}